### PR TITLE
Don't ever allow `~` as a startingDirectory

### LINKED ISSUE #12353: As of latest preview version (1.13.10336.0), homedir alias (~) no longer recognized.
### Windows Terminal version

1.13.10336.0

### Windows build number

22000.466

### Other Software

_No response_

### Steps to reproduce

Open up Terminal Preview, drop down arrow to start a WSL distro session.

### Expected Behavior

Session would open normally.

### Actual Behavior

[error 2147942667 (0x8007010b) when launching `ubuntu.exe']
Could not access starting directory "~"

### PR DESCRIPTION
**Pull request by zadjii-msft**
Basically, some WSL distros ship fragments that replace the `commandline` with the executable for their distro (`ubuntu.exe`, etc.). We didn't expect that when we changed the `startingDirectory` for them all to `~`. 

Unfortunately, `~` is really never a valid path for a process on windows, so those distros would now fail with

```
[error 2147942667 (0x8007010b) when launching `ubuntu1804.exe']
Could not access starting directory "~"
```

If we find that we were unable to mangle `~` into the user's WSL `commandline`, then we will re-evaluate that `startingDirectory` as `%USERPROFILE%`, which is at least something sensible, if albeit not what they wanted. 

* regressed in #12315
* [x] Closes #12353
* [x] Tested with a (`ubuntu1804.exe`, `~`) profile - launched successfully, where 1.13 in market fails.
* [x] added tests

### CODE REVIEW

File: src/types/utils.cpp [hunk 1]
```diff
@@ -670,8 +672,15 @@ std::tuple<std::wstring, std::wstring> Utils::MangleStartingDirectoryForWSL(std:
...
+    // GH #12353: `~` is never a valid windows path. We can only accept that as
+    // a startingDirectory when the exe is specifically wsl.exe, because that
+    // can override the real startingDirectory. If the user set the
+    // startingDirectory to ~, but the commandline to something like pwsh.exe,
+    // that won't actually work. In that case, mangle the startingDirectory to
+    // %userprofile%, so it's at least something reasonable.
     return {
         std::wstring{ commandLine },
-        std::wstring{ startingDirectory }
+        std::wstring{ startingDirectory == L"~" ? wil::ExpandEnvironmentStringsW<std::wstring>(L"%USERPROFILE%") :
```
Comment (DHowett) [inline src/types/utils.cpp:683]: can we just return a null startingdirectory and let the OS take care of it? or should we always resolve userprofile?

Comment (DHowett) [inline src/types/utils.cpp:683]: you are not using the `executableWasWSL` var from earlier

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/types/utils.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/types/utils.cpp:683]: We certainly could. I thought `userprofile` would be closer to what users expect, than... system32? Whatever the terminal's cwd is? `ubuntu.exe` ignores it's CWD, but I don't know about other launchers, so I dunno.... pwsh would use the startingDirectory I guess

Comment (zadjii-msft) [reply src/types/utils.cpp:683]: Okay so replacing the starting directory with `""` then gets treated as "Use Parent Process Directory". (which is basically the same as... `.`? )

IMO the userprofile one is more predictable, but overall I don't care one way or the other.

Comment (zadjii-msft) [reply src/types/utils.cpp:683]: I pushed that change to [`dev/migrie/b/12353-with-null`](https://github.com/microsoft/terminal/compare/dev/migrie/b/12353-with-null), if we make a decision and need it merged tonight.

Comment (DHowett) [reply src/types/utils.cpp:683]: fine w userprofile. should we strip the unused variable?

File: src/types/utils.cpp [hunk 2]
```diff
@@ -600,6 +600,7 @@ bool Utils::IsElevated()
 std::tuple<std::wstring, std::wstring> Utils::MangleStartingDirectoryForWSL(std::wstring_view commandLine,
                                                                             std::wstring_view startingDirectory)
 {
+    bool executableWasWSL{ false };
```
Comment (DHowett) [inline src/types/utils.cpp:603]:
```suggestion
```

File: src/types/utils.cpp [hunk 3]
```diff
@@ -611,6 +612,7 @@ std::tuple<std::wstring, std::wstring> Utils::MangleStartingDirectoryForWSL(std:
             const auto executableFilename{ executablePath.filename().wstring() };
             if (executableFilename == L"wsl" || executableFilename == L"wsl.exe")
             {
+                executableWasWSL = true;
```
Comment (DHowett) [inline src/types/utils.cpp:615]:
```suggestion
```

File: src/types/utils.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/types/utils.cpp:683]: > fine w userprofile. should we strip the unused variable?

butts that was supposed to be pushed

**Review by DHowett — APPROVED** [review]

File: src/types/utils.cpp [hunk 4]
```diff
@@ -670,8 +670,15 @@ std::tuple<std::wstring, std::wstring> Utils::MangleStartingDirectoryForWSL(std:
...
+    // a startingDirectory when the exe is specifically wsl.exe, because that
+    // can override the real startingDirectory. If the user set the
+    // startingDirectory to ~, but the commandline to something like pwsh.exe,
+    // that won't actually work. In that case, mangle the startingDirectory to
+    // %userprofile%, so it's at least something reasonable.
     return {
         std::wstring{ commandLine },
-        std::wstring{ startingDirectory }
+        std::wstring{ startingDirectory == L"~" ? wil::ExpandEnvironmentStringsW<std::wstring>(L"%USERPROFILE%") :
+                                                  startingDirectory }
```
Comment (DHowett) [inline src/types/utils.cpp:682]:
```suggestion
        startingDirectory == L"~" ? wil::ExpandEnvironmentStringsW<std::wstring>(L"%USERPROFILE%") :
                                    std::wstring{ startingDirectory }
```

Comment (DHowett) [reply src/types/utils.cpp:682]: fixes SA!

**Comment by DHowett** [comment]
format is clean

**Review by miniksa — APPROVED** [review]

**Comment by DHowett** [comment]
@msftbot merge this in 2 minutes

**Pull request merged**